Python/compile.c

```diff
@@ -2065,57 +2063,47 @@ static int
 compiler_visit_annotations(struct compiler *c, arguments_ty args,
                            expr_ty returns)
 {
-    /* Push arg annotation dict.
+    /* Push arg annotation names and values.
        The expressions are evaluated out-of-order wrt the source code.
 
-       Return 0 on error, -1 if no dict pushed, 1 if a dict is pushed.
+       Return 0 on error, -1 if no annotations pushed, 1 if a annotations is pushed.
        */
     static identifier return_str;
-    PyObject *names;
-    Py_ssize_t len;
-    names = PyList_New(0);
-    if (!names)
-        return 0;
+    Py_ssize_t annotations_len = 0;
 
-    if (!compiler_visit_argannotations(c, args->args, names))
+    if (!compiler_visit_argannotations(c, args->args, &annotations_len))
         goto error;
-    if (!compiler_visit_argannotations(c, args->posonlyargs, names))
+    if (!compiler_visit_argannotations(c, args->posonlyargs, &annotations_len))
         goto error;
     if (args->vararg && args->vararg->annotation &&
         !compiler_visit_argannotation(c, args->vararg->arg,
-                                     args->vararg->annotation, names))
+                                     args->vararg->annotation, &annotations_len))
         goto error;
-    if (!compiler_visit_argannotations(c, args->kwonlyargs, names))
+    if (!compiler_visit_argannotations(c, args->kwonlyargs, &annotations_len))
         goto error;
     if (args->kwarg && args->kwarg->annotation &&
         !compiler_visit_argannotation(c, args->kwarg->arg,
-                                     args->kwarg->annotation, names))
+                                     args->kwarg->annotation, &annotations_len))
         goto error;
 
     if (!return_str) {
         return_str = PyUnicode_InternFromString("return");
         if (!return_str)
             goto error;
     }
-    if (!compiler_visit_argannotation(c, return_str, returns, names)) {
+    if (!compiler_visit_argannotation(c, return_str, returns, &annotations_len)) {
         goto error;
     }
 
-    len = PyList_GET_SIZE(names);
-    if (len) {
-        PyObject *keytuple = PyList_AsTuple(names);
-        Py_DECREF(names);
-        ADDOP_LOAD_CONST_NEW(c, keytuple);
-        ADDOP_I(c, BUILD_CONST_KEY_MAP, len);
+    if (annotations_len) {
+        ADDOP_I(c, BUILD_TUPLE, annotations_len);
         return 1;
     }
     else {
-        Py_DECREF(names);
         return -1;
     }
 
 error:
-    Py_DECREF(names);
     return 0;
```

Comment:
Can we just `return 0` instead of `goto error;`?

Reply:
Sure, fixed